src/system/random/basic.lean

```diff
@@ -0,0 +1,467 @@
+      apply mod_lt,
+      rw add_one, apply zero_lt_succ },
+    { transitivity x' + (y' - x'),
+      apply le_of_eq, ac_refl,
+      rw [← nat.add_sub_assoc P,nat.add_sub_cancel_left], },
+    simp only [add_comm _ x',add_comm _ 1],
+    apply bitvec.interval_fits_in_word_size P,
+  end,
+⟨ bitvec.of_nat _ r , Hx , Hy ⟩
+
+instance random_bitvec (n : ℕ) : random (bitvec n) :=
+{ random := λ _ inst, @bitvec.random _ inst n,
+  random_r := λ _ inst x y p, bitvec.fit_in_Icc _ _ p <$> @bitvec.random _ inst n }
+
+open nat
+
+/-- `word_size n` gives us the number of bits required to represent `n` -/
+def word_size (x : ℕ) : ℕ :=
+log 2 x + 1
+
+local infix ^ := nat.pow
+
+namespace fin
+section fin
+parameters {n : ℕ} [fact (0 < n)]
+
+/-- `random_aux m k` `m` words worth of random numbers and combine them
+with `k` -/
+protected def random_aux : ℕ → ℕ → rand_g g (fin n)
+| 0 k := return $ fin.of_nat' k
+| (succ n) k :=
+do x ← rand_g.next,
+   random_aux n $ x + (k * shift_31_left)
+
+/-- generate a `fin` randomly -/
+protected def random : rand_g g (fin n) :=
+let m := word_size n / 31 + 1 in
+random_aux m 0
+
+section fit_in_Icc
+
+parameters {i' r k : ℕ}
+parameters {y : fin k}
+
+parameters {x' : ℕ}
+
+parameters P' : x' ≤ y.val
+include P'
+
+lemma interval_fits_in_word_size : x' + i' % (1 + (y.val - x')) < k :=
+begin
+  apply @lt_of_lt_of_le _ _ _ (x' + (y.val - x' + 1)),
+  { apply add_lt_add_left, simp only [add_comm 1],
+    apply nat.mod_lt, apply zero_lt_succ },
+  { rw [← add_assoc,← nat.add_sub_assoc P',nat.add_sub_cancel_left,add_one],
+    apply y.is_lt }
+end
+
+end fit_in_Icc
+
+/-- Use `i` to generate an element of the interval `x .. y` -/
+protected def fit_in_Icc {n : ℕ} (x y : fin n) (P : x ≤ y)
+  (i : fin n) : (x .. y) :=
+let x' := x.val,
+    i' := i.val,
+    r := i' % (y.val - x' + 1) + x' in
+have P' : x.val ≤ y.val,
+  by { rw ← le_def, apply P },
+by haveI : fact (0 < n) :=  lt_of_le_of_lt (nat.zero_le x.val) x.is_lt; exact
+have Hx : x ≤ fin.of_nat' r,
+  begin
+    dsimp [r],
+    simp only [fin.le_def,fin.val_of_nat_eq_mod',add_comm _ x',add_comm _ 1],
+    rw [mod_eq_of_lt],
+    { apply nat.le_add_right },
+    apply fin.interval_fits_in_word_size,
+    dsimp [x'],
+    rw ← fin.le_def, apply P
+  end,
+have Hy : fin.of_nat' r ≤ y,
+  begin
+    dsimp [r],
+    rw [fin.le_def,fin.val_of_nat_eq_mod',mod_eq_of_lt],
+    transitivity (y.val - x') + x',
+    { apply add_le_add_right,
+      apply le_of_lt_succ,
+      rw add_one,
+      apply mod_lt,
+      apply zero_lt_succ },
+    { transitivity x' + (y.val - x'),
+      apply le_of_eq, ac_refl,
+      rw [← nat.add_sub_assoc P',nat.add_sub_cancel_left], },
+    simp only [add_comm _ x',add_comm _ 1],
+    apply fin.interval_fits_in_word_size P',
+  end,
+⟨ fin.of_nat' r , Hx , Hy ⟩
+
+/-- generate an element of the interval `x .. y` -/
+protected def random_r (x y : fin n) (p : x ≤ y) : rand_g g (x .. y) :=
+fin.fit_in_Icc _ _ p <$> fin.random
```

Comment:
As for `bitvec` and `fin.random`, the output is not uniformly distributed if `n` is not a power of `2` or if `y - x + 1` does not divide `n`.